app/controllers/api/v1/users_controller.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+module API
+  module V1
+    class UsersController < API::V1::ApplicationController
+      include API::V1::ErrorHandlerConcern
+
+      skip_before_action :doorkeeper_authorize!, only: :create
+
+      def create
+        user = User.new(create_params.except(:client_id))
+        client_app = Doorkeeper::Application.find_by(uid: create_params[:client_id])
+
+        unless client_app
+          render_error 'Invalid client ID', status: :forbidden, source: :client_id
+          return
+        end
+
+        render_errors ActiveModel::ErrorsSerializer.new(user.errors).serializable_hash and return unless user.save
+
+        # create access token for the user, so the user won't need to login again after registration
+        access_token = user.get_access_token client_app.id
+        # return json containing access token and refresh token
+        # so that user won't need to call login API right after registration
+        render json: create_success_result(user, access_token), status: :created
+      end
+
+      private
+
+      def create_success_result(user, access_token)
+        user_json = UserSerializer.new(user).serializable_hash
+        token_json = Doorkeeper::TokenSerializer.new(access_token).serializable_hash
```

Comment:
@malparty How about remove this method entirely by creating a new serializer, by having token as a part of user.

```
    UserTokenSerializer < UserSerializer
       attribute :access_token

        def access_token
              @access_token ||= object.create_access_token
       end
end

Reply:
It would change the result signature (token being nested inside the user object -- is it okay for JsonApi? If I get it right it only prevent us to specify an explicit relationship between the user and his token), but I like the "cleanness" of this idea!!! :) 

I will give it a try before monday!! :) // Need to go now =/

Reply:
okay, I've submited a proposal. 
It's not as clean as your solution because we need to pass the ClientId in order to create the new token on the fly. 
A simplest solution used in most of the other IC is to NOT send back a token when the user just signed in. But as the purpose is to learn, I think it's great to dive into this :)

Reply:
@ankitkalia1195   @junan  feel free to comments in here 
(Junan, I will resolve the other comment quite related to this thread)